client/hudwidget.cpp

```diff
@@ -1731,3 +1770,135 @@ void hud_battle_log::showEvent(QShowEvent *event)
   m_timer.restart();
   setVisible(true);
 }
+
+/**
+   This predicate is called by cycle_units and friends to filter
+   relevant units from the iterated units.
+
+   current_unit is the currently selected unit.
+   candidate is the candidate checked for relevance.
+ */
+static bool cycle_units_predicate(const struct unit *current_unit,
+                                  const struct unit *candidate)
+{
+  fc_assert_ret_val(current_unit != nullptr && candidate != nullptr, false);
+
+  if (candidate->utype != current_unit->utype) {
+    return false;
+  }
+
+  if (ACTIVITY_IDLE != candidate->activity
+      && ACTIVITY_SENTRY != candidate->activity) {
+    return false;
+  }
+
+  if (!can_unit_do_activity(candidate, ACTIVITY_IDLE)) {
+    return false;
+  }
+
+  return true;
+}
+
+/**
+  Gets the count of relevant units and determines the index of the
+  current unit among these units.
+
+  If current_unit is not among the relevant units, then index 0 is
+  assumed by default.
+
+  current_unit is the currently selected unit.
+ */
+static std::tuple<int, int>
+cycle_units_get_count_and_index(const struct unit *current_unit)
+{
+  fc_assert_ret_val(current_unit != nullptr, std::make_tuple(0, 0));
+
+  int unit_count = 0;
+  int current_unit_index = 0;
+
+  unit_list_iterate(client_player()->units, punit)
+  {
+    if (!cycle_units_predicate(current_unit, punit)) {
+      continue;
+    }
+
+    if (current_unit->id == punit->id) {
+      current_unit_index = unit_count;
+    }
+
+    unit_count++;
+  }
+  unit_list_iterate_end;
+
+  return {unit_count, current_unit_index};
+}
+
+/**
+   Selects the next unit when units are cycled.
+
+   current_unit is the currently selected unit.
+   index is the index of the unit to be selected next among the
+   relevant units.
+ */
+static void cycle_units_select_index(const struct unit *current_unit,
+                                     int index)
+{
+  fc_assert_ret(current_unit != nullptr);
+
+  // Select the next unit.
+  int unit_count = 0;
+  unit_list_iterate(client_player()->units, punit)
+  {
+    if (!cycle_units_predicate(current_unit, punit)) {
+      continue;
+    }
+
+    if (unit_count == index) {
+      unit_focus_set_and_select(punit);
+    }
+
+    unit_count++;
+  }
+  unit_list_iterate_end;
+}
+
+/**
+   Cycle idle or sentried units for the unittype of the currently
+   selected unit.
+
+   The sign of direction determines in which direction the units will
+   be cycled.
+ */
+void cycle_units(const int direction)
+{
+  if (get_num_units_in_focus() != 1) {
+    return;
+  }
+
+  struct unit *current_unit = head_of_units_in_focus();
+
+  int unit_count = 0;
+  int current_unit_index = 0;
+  std::tie(unit_count, current_unit_index) =
```

Comment:
Structured binding (we have C++17, let's make use of it!)
```suggestion
  auto [unit_count, current_unit_index] =
```